### ISSUE TITLE
HybPhyloMaker6a_RAxML_for_selected errors

### ISSUE THREAD

**Issue by biomendi** [tomas-fer/HybPhyloMaker#10]
I have noticed that the script 'HybPhyloMaker6a_RAxML_for_selected' will show errors for some of my alignments, as follows: 
```
HybPhyloMaker6a is running locally...
Working with exons...with corrected reading frame

Testing if input data are available...OK


Generating RAxML trees with 100 rapid bootstrap replicates using GTRGAMMA model...

Modifying selected FASTA files...

Generating RAxML trees...
Processing file: CorrectedAssembly_10005_modif50 (1 out of 698)
Reduced alignment not found...using original alignment
Processing file: CorrectedAssembly_10010_modif50 (2 out of 698)
Reduced alignment found...using it
Processing file: CorrectedAssembly_10011_modif50 (3 out of 698)
Reduced alignment found...using it
.
.
.
Processing file: CorrectedAssembly_10063_modif50 (7 out of 698)
Reduced alignment not found...using original alignment
cp: cannot stat '*CorrectedAssembly_10063_modif50.result': No such file or directory
.
.
.
Processing file: CorrectedAssembly_9988_modif50 (697 out of 698)
Reduced alignment not found...using original alignment
cp: cannot stat '*CorrectedAssembly_9988_modif50.result': No such file or directory
Processing file: CorrectedAssembly_9997_modif50 (698 out of 698)
Reduced alignment not found...using original alignment
cp: cannot stat '*CorrectedAssembly_9997_modif50.result': No such file or directory

HybPhyloMaker 6a finished...

Run script HybPhyloMaker6a2 in order to calculate tree properties...
```
After exploring the log file `exons/82trees_corrected50_50/RAxML/raxml.log` I found that these errors are likely caused by not enough species in the alignments:

```
Processing file: CorrectedAssembly_9997_modif50

RAxML can't, parse the alignment file as phylip file 
it will now try to parse it as FASTA file

TOO FEW SPECIES
Reduced alignment not found...using original alignment

RAxML can't, parse the alignment file as phylip file 
it will now try to parse it as FASTA file

TOO FEW SPECIES
```
I think these errors do not affect the results, right? However, would it maybe be possible to somehow let the user know about how many genes do not have enough species (or the opposite: how many do have enough species)?

**Comment by biomendi** [comment]
HybPhyloMaker6a2 did not work... Maybe because of the errors above?
```
HybPhyloMaker6a2 is running locally...
Working with exons...with corrected reading frame

Testing if input data are available...OK


Calculating tree properties...

Calculating and parsing LB score...

paste: tree_stats_table.csv: No such file or directory
Plotting boxplots/histograms for tree properties...

cp: *.png: No such file or directory
Combining alignment and tree properties...

Plotting gene properties correlations for RAxML trees...

cp: genes_corrs.*: No such file or directory
rm: genes_corrs.*: No such file or directory
HybPhyloMaker 6a2 finished...

```

**Comment by biomendi** [comment]
Please find attached the `exons/82trees_corrected50_50/R.log` produced by HybPhyloMaker6a2

```
.
.
.
Warning message:
In cbind(paste(aln_dir, aln_files, sep = ""), paste(trees_dir, trees_files,  :
  number of rows of result is not a multiple of vector length (arg 2)
Error in model.frame.default(formula = p_dist ~ branch_dist, drop.unused.levels = TRUE) : 
  variable lengths differ (found for 'branch_dist')
Calls: t ... lm -> eval -> eval -> <Anonymous> -> model.frame.default
Execution halted

Calculating and parsing LB score using LBscores.R


Attaching package: ‘seqinr’

The following objects are masked from ‘package:ape’:

    as.alignment, consensus

There were 50 or more warnings (use warnings() to see the first 50)

Plotting boxplots/histograms for tree properties using treepropsPlot.r

Error in read.table(file = file, header = header, sep = sep, quote = quote,  : 
  no lines available in input
Calls: read.csv -> read.table
Execution halted

Plotting gene properties correlations for RAxML trees using plotting_correlations.R

Error in eval(substitute(select), nl, parent.frame()) : 
  object 'Bootstrap' not found
Calls: subset -> subset.data.frame -> eval -> eval
Execution halted
```

**Comment by tomas-fer** [comment]
Yes, RAxML does not work for alignments with less than 4 samples. And 6a2 also does not work properly because there is not an equal number of alignments and trees.
There is a new version of 5NEW script which deletes all alignments with less than 4 samples. This creates a list of selected genes which should well go through 6a etc. Hope this helps.

**Comment by biomendi** [comment]
I am re-running 5NEW script right now. 

Regarding: https://github.com/tomas-fer/HybPhyloMaker/issues/9#issuecomment-446990613 - Do you have any idea about the issue with the script 4b?

Thanks a lot!

**Comment by biomendi** [comment]
Finished re-running 5NEW, 6a and 6a2. 

Here the output of 6a2:

```
HybPhyloMaker6a2 is running locally...
Working with exons...with corrected reading frame

Testing if input data are available...OK


Calculating tree properties...

Calculating and parsing LB score...

Plotting boxplots/histograms for tree properties...

Combining alignment and tree properties...

Plotting gene properties correlations for RAxML trees...

cp: genes_corrs.*: No such file or directory
rm: genes_corrs.*: No such file or directory
HybPhyloMaker 6a2 finished...

```
And the R.log 
```
.
.
.
Calculating and parsing LB score using LBscores.R


Attaching package: ‘seqinr’

The following objects are masked from ‘package:ape’:

    as.alignment, consensus

There were 50 or more warnings (use warnings() to see the first 50)

Plotting boxplots/histograms for tree properties using treepropsPlot.r

Error in x[floor(d)] + x[ceiling(d)] : 
  non-numeric argument to binary operator
Calls: boxplot -> boxplot.default -> boxplot.stats -> <Anonymous>
Execution halted

Plotting gene properties correlations for RAxML trees using plotting_correlations.R

Error in read.table("combined.txt", header = TRUE) : 
  more columns than column names
Execution halted

```

**Comment by tomas-fer** [comment]
For the script 6a2, it is crucial that there is the same number of loci 'selected_genes.txt' and the number of trees generated by RAxML. This seems to be corrected in the script 5NEW previously. But check it first, please...

**Comment by tomas-fer** [comment]
Also, I want to see 'tree_stats_table.csv' and 'LBscoresSDPerLocus.txt'. Thanks...

**Comment by biomendi** [comment]
Thanks for your help

It seems that I have the same number of loci and gene trees as generated by RAxML:

`wc -l 81selected_corrected50/selected_genes_50_50.txt` --> 256

`ls 82trees_corrected50_50/RAxML/RAxML_bipartitionsBranchLabels.CorrectedAssembly_* | wc -l` --> 256

`wc -l tree_stats_table.csv` --> 257 (first row is the header)

I attach the [tree_stats_table.csv.txt](https://github.com/tomas-fer/HybPhyloMaker/files/2686412/tree_stats_table.csv.txt)

**Comment by tomas-fer** [comment]
There is a problem when computing 'tree_stats_table.csv.txt'. I updated 'tree_props.r' in 'HybSeqSource' folder. Could you try it now?

**Comment by biomendi** [comment]
Script 6a2 works fine now — While re-running the scripts, I have noticed a "No such file or directory" error on script 6a that before I did not detect (but it was also there). 

The error occurs only for few genes, below one example: 

```
.
.
.
Processing file: CorrectedAssembly_9893_modif50 (254 out of 256)
Reduced alignment found...using it
mv: cannot stat 'CorrectedAssembly_9893.part.reduced': No such file or directory
Processing file: CorrectedAssembly_9909_modif50 (255 out of 256)
Reduced alignment found...using it
Processing file: CorrectedAssembly_9923_modif50 (256 out of 256)
Reduced alignment not found...using original alignment

HybPhyloMaker 6a finished...

Run script HybPhyloMaker6a2 in order to calculate tree properties...
```

**Comment by biomendi** [comment]
Also, if I am not mistaken, the script `HybPhyloMaker8e_concatenatedFastTree.sh` likely has the same "Argument list too long" error as the script `HybPhyloMaker8f_concatenatedExaML.sh` : https://github.com/tomas-fer/HybPhyloMaker/issues/9#issuecomment-447405067

**Comment by biomendi** [comment]
These are two examples:

```
Processing file: CorrectedAssembly_8085_modif50

RAxML can't, parse the alignment file as phylip file 
it will now try to parse it as FASTA file



IMPORTANT WARNING: Sequences Oreobates-cruralis_RS21 and Oreobates-cruralis_RS22 are exactly identical

IMPORTANT WARNING
Found 1 sequence that is exactly identical to other sequences in the alignment.
Normally they should be excluded from the analysis.


A mixed model file with model assignments for undetermined
columns removed has already been printed to  file CorrectedAssembly_8085.part.reduced
Just in case you might need it, an alignment file with 
sequence duplicates removed is printed to file CorrectedAssembly_8085_modif50.fas.reduced


This is RAxML version 8.2.10 released by Alexandros Stamatakis on March 2017.
.
.
.

```

```
Processing file: CorrectedAssembly_9893_modif50

RAxML can't, parse the alignment file as phylip file 
it will now try to parse it as FASTA file



IMPORTANT WARNING: Sequences Oreobates-cruralis_RS16 and Oreobates-cruralis_RS17 are exactly identical


IMPORTANT WARNING: Sequences Oreobates-cruralis_RS16 and Oreobates-cruralis_RS21 are exactly identical


IMPORTANT WARNING: Sequences Oreobates-cruralis_RS16 and Oreobates-cruralis_RS22 are exactly identical

IMPORTANT WARNING
Found 3 sequences that are exactly identical to other sequences in the alignment.
Normally they should be excluded from the analysis.


A mixed model file with model assignments for undetermined
columns removed has already been printed to  file CorrectedAssembly_9893.part.reduced
Just in case you might need it, an alignment file with 
sequence duplicates removed is printed to file CorrectedAssembly_9893_modif50.fas.reduced


This is RAxML version 8.2.10 released by Alexandros Stamatakis on March 2017.
```

**Comment by biomendi** [comment]
I found "Argument list too long" error in `HybPhyloMaker_reports.sh`

```
HybPhyloMaker-summary is running locally...
Working with exons...with corrected reading frame


Collecting tables...
Modifying tables...
HybPhyloMaker_reports.sh: line 241: /usr/local/opt/coreutils/libexec/gnubin/ls: Argument list too long
Writing XLSX...
ARGUMENT '5' __ignored__

ARGUMENT '0' __ignored__

ARGUMENT '256' __ignored__

ARGUMENT '50' __ignored__

[1] "Nr. samples is 5"
[1] "Nr. genes is 0"
[1] "Nr. selected genes is 256"
[1] "Missing percent level is 50"
[1] "Reading tables"
[1] "Making list"
[1] "Defining styles"
[1] "Saving workbook"
[1] "Deleting empty cells"
[1] "Adding styles"
[1] "...sheet1"
[1] "...sheet2"
[1] "...sheet3"
[1] "...sheet4"
[1] "...sheet5"
[1] "...sheet6"
[1] "Setting column width"
[1] "Freezing header"
[1] "Adding formulas"
[1] "Writing final file"

Script HybPhyloMaker-summary finished...
```

**Comment by biomendi** [comment]
Also, it will it be great to check if the `'No such file or directory'` error in the output of the script `HybPhyloMaker6a_RAxML_for_selected.sh` does affect the final results or not. I am now re-running the script 6a (which will take 2 days) and it would be very helpful to know this information in order to save some time...

**Comment by biomendi** [comment]
Found unnoticed 'Argument list too long' in `HybPhyloMaker6a2_RAxML_trees_summary.sh`

```
HybPhyloMaker6a2 is running locally...
Working with exons...with corrected reading frame

Testing if input data are available...OK

HybPhyloMaker6a2_RAxML_trees_summary.sh: line 154: /usr/local/opt/coreutils/libexec/gnubin/cp: Argument list too long
mv: cannot stat '*': No such file or directory

Calculating tree properties...

Calculating and parsing LB score...

paste: tree_stats_table.csv: No such file or directory
Plotting boxplots/histograms for tree properties...

cp: cannot stat '*.png': No such file or directory
Combining alignment and tree properties...

Plotting gene properties correlations for RAxML trees...

cp: cannot stat 'genes_corrs.*': No such file or directory
rm: cannot remove 'genes_corrs.*': No such file or directory
HybPhyloMaker 6a2 finished...

```
Line 154: 
```
#Copy all RAxML tree files (*.tre) with bootstrap values to subfolder 'trees'
cp $path/${treepath}${MISSINGPERCENT}_${SPECIESPRESENCE}/RAxML/*bipartitions.*Assembly* trees/
```

**Comment by tomas-fer** [comment]
@biomendi, could you send me some of the 'problematic' alignments that are causing "No such file or directory" error when running 'HybPhyloMaker6a_RAxML_for_selected.sh'? I need to reproduce the error to be able to fix it. Thanks.

**Comment by biomendi** [comment]
> This option is also necessary for the script 5NEW to work.

The script 5NEW worked previously with the old "settings.cfg" and the script HybPhyloMaker6a took three days to finish (with the above mentioned "No such file or directory" errors). I was able to fix HybPhyloMaker6a2, so I obtained the astral/astrid/mrl trees. Now I was just missing the concatenated trees. Do you recommend re-running from 5NEW with the new settings.cfg?

**Comment by tomas-fer** [comment]
No, there is no need to re-run the 5NEW script, but use the new settings in the future (there is now one new option 'AMAS' which is active in 6a2 and 8e: fast=the original solution, slow=a workaround for the case of 'Argument list too long'). Until there will be a fix by the author of AMAS my solution is slow and I decided to keep the original solution too (for the cases where 'Argument list too long' is not an issue).
I fixed 6a2 differently but with the same effect, so should get the same species trees with 8a-e scripts.
